src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+              assume_false_location, goto_programt::make_skip());
+            for(auto const &restriction_target : restriction.second)
+            {
+              auto new_instruction = original_function_call_instruction;
+              // can't use get_function_call because that'll return a const ref
+              const symbol_exprt &function_pointer_target_symbol_expr =
+                goto_model.symbol_table.lookup_ref(restriction_target)
+                  .symbol_expr();
+              to_code_function_call(new_instruction.code).function() =
+                function_pointer_target_symbol_expr;
+              auto const goto_end_if_location =
+                goto_function_body.insert_before(
+                  else_location,
+                  goto_programt::make_goto(
+                    end_if_location,
+                    original_function_call_instruction.source_location));
+              auto const replaced_instruction_location =
+                goto_function.second.body.insert_before(
+                  goto_end_if_location, new_instruction);
+              else_location = goto_function.second.body.insert_before(
+                replaced_instruction_location,
+                goto_programt::make_goto(
+                  else_location,
+                  notequal_exprt{
+                    pointer_symbol,
+                    address_of_exprt{function_pointer_target_symbol_expr}}));
+            }
+          }
+        }
+      });
+  }
+}
+
+void parse_function_pointer_restriction_options_from_cmdline(
+  const cmdlinet &cmdline,
+  optionst &options)
+{
+  options.set_option(
+    RESTRICT_FUNCTION_POINTER_OPT,
+    cmdline.get_values(RESTRICT_FUNCTION_POINTER_OPT));
+  options.set_option(
+    RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT,
+    cmdline.get_values(RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT));
+}
+
+namespace
+{
+function_pointer_restrictionst::restrictionst
+merge_function_pointer_restrictions(
+  const function_pointer_restrictionst::restrictionst &lhs,
+  const function_pointer_restrictionst::restrictionst &rhs)
+{
+  auto result = lhs;
+  for(auto const &restriction : rhs)
+  {
+    auto emplace_result = result.emplace(restriction.first, restriction.second);
+    if(!emplace_result.second)
+    {
+      for(auto const &target : restriction.second)
+      {
+        emplace_result.first->second.insert(target);
+      }
+    }
+  }
+  return result;
+}
+
+function_pointer_restrictionst::restrictionst
+parse_function_pointer_restrictions_from_command_line(
+  const std::list<std::string> &restriction_opts)
+{
+  auto function_pointer_restrictions =
+    function_pointer_restrictionst::restrictionst{};
+  for(auto const &restriction_opt : restriction_opts)
+  {
+    // the format for restrictions is <pointer_name>/<target[,more_targets]*>
+    // exactly one pointer and at least one target
+    auto const pointer_name_end = restriction_opt.find('/');
```

Comment:
:question: Could regex be used here? Rather than manually parsing the string

Reply:
It could be, though it probably wouldn't improve things much. The current version has the advantage that specific error messages are printed for different format errors.